subprojects/core/src/main/java/org/gradle/api/internal/TaskExecutionHistory.java

```diff
@@ -15,15 +15,16 @@
  */
 package org.gradle.api.internal;
 
-import org.gradle.api.file.FileCollection;
-
 import javax.annotation.Nullable;
+import java.io.File;
+import java.util.Set;
 
 public interface TaskExecutionHistory {
     /**
      * Returns the set of output files which the task produced.
      */
-    FileCollection getOutputFiles();
+    @Nullable
```

Comment:
We don't even check in `DefaultTaskOutputs.getPreviousOutputFiles()` whether the return value is `null` or not, we return it as non-nullable, and use it as such in `SimpleStaleClassCleaner`.

It would be nice to make sure we never return `null` here. We could add a `hasOutputHistory()` method or something that we can check in `SkipEmptySourceFilesTaskExecuter`.